File: examples/typescript/ethereum.ts
import { createPublicClient, http, formatUnits } from 'viem';
import { sepolia } from 'viem/chains';
import config from '../../config.json' assert { type: 'json' };

// ERC-20 ABI (just the balanceOf function)
const erc20Abi = [
    {
        inputs: [{ name: 'account', type: 'address' }],
        name: 'balanceOf',
        outputs: [{ name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
] as const;

interface Balance {
    token: string;
    amount: string;
    decimals: number;
    formatted: string;
}

async function getSepoliaBalances(address: `0x${string}`): Promise<Balance[]> {
    const chainConfig = config.chains.sepolia;

    const client = createPublicClient({
        chain: sepolia,
        transport: http(chainConfig.rpc),
    });

    const balances: Balance[] = [];

    // Get ETH balance
    const ethBalance = await client.getBalance({ address });
    balances.push({
        token: chainConfig.nativeToken.symbol!,
        amount: ethBalance.toString(),
        decimals: chainConfig.nativeToken.decimals,
        formatted: formatUnits(ethBalance, chainConfig.nativeToken.decimals),
    });

    // Get token balances
    for (const [symbol, tokenInfo] of Object.entries(chainConfig.tokens)) {
        const tokenBalance = await client.readContract({
            address: tokenInfo.address as `0x${string}`,
            abi: erc20Abi,
            functionName: 'balanceOf',
            args: [address],
        });

        balances.push({
            token: symbol,
            amount: tokenBalance.toString(),
            decimals: tokenInfo.decimals,
            formatted: formatUnits(tokenBalance, tokenInfo.decimals),
        });
    }

    return balances;
}

// Example usage
const address = '0x78697a9cfc48c1e9d1040172d51833ef78083b10';
const balances = await getSepoliaBalances(address as `0x${string}`);

console.log(`\nChain: ${config.chains.sepolia.name}`);
console.log('='.repeat(60));
balances.forEach((balance) => {
    console.log(`${balance.token.padEnd(6)} | ${balance.formatted.padStart(20)} (raw: ${balance.amount})`);
});
console.log('='.repeat(60));

File: src/ethereum.rs
use alloy::primitives::{Address, U256};
use alloy::providers::{Provider, ProviderBuilder};
use alloy::rpc::types::BlockNumberOrTag;
use alloy::sol;
use anyhow::Result;
use async_trait::async_trait;

use crate::chain::ChainProvider;
use crate::types::{Balance, Token};

// ERC-20 ABI for balanceOf
sol! {
    #[sol(rpc)]
    interface IERC20 {
        function balanceOf(address account) external view returns (uint256);
    }
}

/// Ethereum chain provider using JSON-RPC
pub struct EthereumProvider {
    rpc_url: String,
}

impl EthereumProvider {
    pub fn new(rpc_url: String) -> Self {
        Self { rpc_url }
    }

    pub fn new_sepolia() -> Self {
        Self {
            // Using public Sepolia RPC endpoint
            rpc_url: "https://ethereum-sepolia-rpc.publicnode.com".to_string(),
        }
    }
}

#[async_trait]
impl ChainProvider for EthereumProvider {
    async fn get_native_balance(&self, address: &str) -> Result<Balance> {
        let provider = ProviderBuilder::new().on_http(self.rpc_url.parse()?);

        let addr: Address = address.parse()?;
        let balance = provider
            .get_balance(addr)
            .block_id(BlockNumberOrTag::Latest.into())
            .await?;

        Ok(Balance::new("ETH".to_string(), balance.to_string(), 18))
    }

    async fn get_token_balance(&self, address: &str, token: &Token) -> Result<Balance> {
        let Token::Erc20 {
            address: token_address,
            symbol,
            decimals,
        } = token;

        let provider = ProviderBuilder::new().on_http(self.rpc_url.parse()?);

        let addr: Address = address.parse()?;
        let token_addr: Address = token_address.parse()?;

        let contract = IERC20::new(token_addr, provider);
        let balance: U256 = contract.balanceOf(addr).call().await?._0;

        Ok(Balance::new(symbol.clone(), balance.to_string(), *decimals))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::chain::ChainProvider;
    use crate::types::Token;

    #[tokio::test]
    #[ignore] // Requires network access
    async fn test_sepolia_specific_address_balances() {
        let provider = EthereumProvider::new_sepolia();
        let address = "0x78697a9cfc48C1e9d1040172d51833EF78083b10";

        // Check ETH Balance > 0
        let eth_balance = provider.get_native_balance(address).await.unwrap();
        assert_eq!(eth_balance.token, "ETH");
        let eth_amount: f64 = eth_balance.formatted.parse().unwrap();
        assert!(
            eth_amount > 0.0,
            "ETH balance {} should be > 0.007",
            eth_amount
        );

        // Define expected tokens
        let usdc = Token::Erc20 {
            address: "0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238".to_string(),
            symbol: "USDC".to_string(),
            decimals: 6,
        };

        // Check USDC Balance >= 0.1
        let usdc_balance = provider.get_token_balance(address, &usdc).await.unwrap();
        assert_eq!(usdc_balance.token, "USDC");
        let usdc_amount: f64 = usdc_balance.formatted.parse().unwrap();
        assert!(
            usdc_amount >= 0.1,
            "USDC balance {} should be >= 0.1",
            usdc_amount
        );
    }
}
